### PR TITLE
CDAP-2255 Port TimePartitionedFileSetDatasetAvroSink to use plugin framework

### PR DESCRIPTION
**Pull request by rohitsinha54**

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 1]
```diff
@@ -40,43 +43,64 @@
...
+import javax.annotation.Nullable;
 
 /**
  * A {@link BatchSink} to write Avro record to {@link TimePartitionedFileSet}
  */
+@Plugin(type = "sink")
+@Name("TimePartitionedFileSetDatasetAvroSink")
+@Description("AVRO Sink with Time Partitioned File Dataset")
 public class TimePartitionedFileSetDatasetAvroSink extends
   BatchSink<GenericRecord, AvroKey<GenericRecord>, NullWritable> {
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Curious, can we make this sink accept StructuredRecord like all other sinks?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 2]
```diff
@@ -40,43 +43,64 @@
...
+    private String tpfsName;
+
+    @Name(Properties.TimePartitionedFileSetDataset.SCHEMA)
+    @Description(SCHEMA_DESC)
+    private String schema;
+
+    @Name(Properties.TimePartitionedFileSetDataset.BASE_PATH)
+    @Description(BASE_PATH_DESC)
+    @Nullable
+    private String basePath;
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Is there any default value for this optional property?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 3]
```diff
@@ -40,43 +43,64 @@
...
     // if the base path is provided then we should try to create the fileset here
-    if (!Strings.isNullOrEmpty(stageConfig.getProperties().get(BASE_PATH))) {
-      String tpfsName = stageConfig.getProperties().get(TPFS_NAME);
+    Map<String, String> properties = stageConfig.getProperties();
+    if (!Strings.isNullOrEmpty(properties.get(Properties.TimePartitionedFileSetDataset.BASE_PATH))) {
+      String tpfsName = properties.get(Properties.TimePartitionedFileSetDataset.TPFS_NAME);
       Preconditions.checkArgument(!Strings.isNullOrEmpty(tpfsName), "TimePartitionedFileSet name must be given.");
       pipelineConfigurer.createDataset(tpfsName, TimePartitionedFileSet.class.getName(), FileSetProperties.builder()
-        .setBasePath(stageConfig.getProperties().get(BASE_PATH))
+        .setBasePath(properties.get(Properties.TimePartitionedFileSetDataset.BASE_PATH))
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: You don't have to use properties. You can have a private variable of type `TimePartitionedFileSetDatasetAvroConfig`. and you can access its fields.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 4]
```diff
@@ -40,43 +43,64 @@
...
         .setInputFormat(AvroKeyInputFormat.class)
         .setOutputFormat(AvroKeyOutputFormat.class)
         .setEnableExploreOnCreate(true)
         .setSerDe("org.apache.hadoop.hive.serde2.avro.AvroSerDe")
         .setExploreInputFormat("org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat")
         .setExploreOutputFormat("org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat")
-        .setTableProperty("avro.schema.literal", (stageConfig.getProperties().get(SCHEMA)))
+        .setTableProperty("avro.schema.literal", (properties.get(Properties.TimePartitionedFileSetDataset.SCHEMA)))
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Same here.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 5]
```diff
@@ -84,11 +108,14 @@ public void configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineC
...
     Map<String, String> runtimeArguments = context.getPluginProperties().getProperties();
+    Map<String, String> properties = context.getPluginProperties().getProperties();
     Map<String, String> sinkArgs = Maps.newHashMap();
     TimePartitionedFileSetArguments.setOutputPartitionTime(sinkArgs, context.getLogicalStartTime());
-    TimePartitionedFileSet sink = context.getDataset(runtimeArguments.get(TPFS_NAME), sinkArgs);
-    context.setOutput(runtimeArguments.get(TPFS_NAME), sink);
-    Schema avroSchema = new Schema.Parser().parse(runtimeArguments.get(SCHEMA));
+    TimePartitionedFileSet sink = context.getDataset(runtimeArguments.get(properties.get(
+      Properties.TimePartitionedFileSetDataset.TPFS_NAME)), sinkArgs);
+    context.setOutput(runtimeArguments.get(properties.get(Properties.TimePartitionedFileSetDataset.TPFS_NAME)), sink);
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Same here

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 6]
```diff
@@ -84,11 +108,14 @@ public void configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineC
...
+    Schema avroSchema = new Schema.Parser().parse(runtimeArguments.get(properties.get(
+      Properties.TimePartitionedFileSetDataset.SCHEMA)));
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Same here

**Comment by gokulavasan** [comment]
Took one pass at it. Will take another look when comments are addressed.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 1 above)
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Didn't we wanted a AvroSink in the first place. If that has change we can but I think that should go in a different PR completely.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 2 above)
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: I don't think there should  be default value for  the basepath. Its needed when user want to create the TPFS in the pipeline.
@albertshau : Can you confirm.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 7]
```diff
@@ -40,43 +43,64 @@
 import org.apache.hadoop.mapreduce.Job;
 
 import java.util.Map;
+import javax.annotation.Nullable;
 
 /**
  * A {@link BatchSink} to write Avro record to {@link TimePartitionedFileSet}
  */
+@Plugin(type = "sink")
+@Name("TimePartitionedFileSetDatasetAvroSink")
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: That's a pretty long name :) Can it be `TPFSAvroSink` and use the description to tell more details?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: So, is it optional or required? If it is required, you shouldn't have `@Nullable` annotation.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 8]
```diff
@@ -40,43 +43,64 @@
...
+      this.schema = schema;
+    }
   }
 
   @Override
   public void configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineConfigurer) {
     // if the base path is provided then we should try to create the fileset here
-    if (!Strings.isNullOrEmpty(stageConfig.getProperties().get(BASE_PATH))) {
-      String tpfsName = stageConfig.getProperties().get(TPFS_NAME);
+    Map<String, String> properties = stageConfig.getProperties();
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Shouldn't use the `getProperties` anymore. We should remove the support of that method. Properties of the plugin should always read from the config field. @gokulavasan Please remove that method from stageConfig.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 2 above)
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Its optional and does not need a default value. Thats my understanding.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 9]
```diff
@@ -40,43 +43,64 @@
...
-    configurer.addProperty(new Property(BASE_PATH, "Optional: The base path for the time partitioned fileset. Should" +
-      "be provided if you want the pipeline to create the fileset.", false));
+  /**
+   * Config for TimePartitionedFileSetDatasetAvroSink
+   */
+  public static class TimePartitionedFileSetDatasetAvroConfig extends PluginConfig {
+
+    @Name(Properties.TimePartitionedFileSetDataset.TPFS_NAME)
+    @Description(TPFS_NAME_DESC)
+    private String tpfsName;
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: If the name the field the same as the annotated one, you don't need the `@Name` annotation. E.g. in this example, just rename the field as `name` (the schema and basePath, you don't need annotation).

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: So what happen if it is not set?

Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: We will not create the TPFS in the pipeline and expect it to exist.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java [hunk 10]
```diff
@@ -40,43 +43,64 @@
...
 public class TimePartitionedFileSetDatasetAvroSink extends
   BatchSink<GenericRecord, AvroKey<GenericRecord>, NullWritable> {
 
-  private static final String TPFS_NAME = "name";
-  private static final String SCHEMA = "schema";
-  private static final String BASE_PATH = "basePath";
+  private static final String SCHEMA_DESC = "The schema of the record";
+  private static final String TPFS_NAME_DESC = "Name of the Time Partitioned FileSet Dataset to which the records " +
+    "have to be written";
+  private static final String BASE_PATH_DESC = "Optional: The base path for the time partitioned fileset. Should " +
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: You don't have to say `Optional` here, since the meta info of the property in the REST endpoint already tells it's optional.

**Comment by chtyim** [comment]
LGTM.

**Comment by rohitsinha54** [comment]
Will Merge after manual testing.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java (hunk 1 above)
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/TimePartitionedFileSetDatasetAvroSink.java]: Sure makes sense.

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-RT10

**Comment by albertshau** [comment]
already reviewed, merging.
